torchtune/training/_distributed.py

```diff
@@ -726,3 +751,112 @@ def prepare_mha_for_tp(
     if is_fusion_model:
         model.decoder = decoder
     return model
+
+
+def _get_sdpa_context() -> (
+    Callable[[Optional[Generator[None, None, None]]], Generator[None, None, None]]
+):
+    """
+    Creates a context manager to confine to flash/efficient/cuDNN attention backends.
+
+    Returns:
+        A context manager function that takes an optional context parallel context.
+    """
+
+    @contextlib.contextmanager
+    def context(cp_context: Union[Generator[None, None, None], None] = None):
+        with contextlib.ExitStack() as stack:
+            if cp_context is not None:
+                stack.enter_context(
+                    sdpa_kernel(
+                        [
+                            SDPBackend.FLASH_ATTENTION,
+                            SDPBackend.EFFICIENT_ATTENTION,
+                            SDPBackend.CUDNN_ATTENTION,
+                        ]
+                    )
+                )
+                stack.enter_context(cp_context)
+
+            yield
+
+    return context
+
+
+def get_context_parallel_context(
+    *,
+    cp_enabled: bool = False,
+    world_mesh: torch.distributed.DeviceMesh,
+    model: TransformerDecoder,
+    model_inputs: list[torch.Tensor],
```

Comment:
This should take a dict, since all our recipes use a nested dict for inputs. Also, you can't just call list(batch.values) since there are nested dicts. Since we don't support multimodal for now, you can just explicitly check that tokens are inside the batch in this function.